### ISSUE TITLE
torch._C._jit_pass_dce removes code which is not dead

### ISSUE THREAD

**Issue by TobyRoseman** [pytorch/pytorch#69989]
### 🐛 Describe the bug

The following code works in torch `1.9.1` but produces incorrect results in `1.10.0`. Code is being removed from the graph which is not dead code.
```python
import torch

class TensorAssignModel(torch.nn.Module):
    def __init__(self):
        super(TensorAssignModel, self).__init__()

    def forward(self, x):
        x[2] = 9
        return x

x = torch.randn(5, 4)
torch_m = TensorAssignModel()
torch_m = torch.jit.trace(torch_m, x.clone())

graph = torch_m.forward.graph
print(graph)    # This looks good.

torch._C._jit_pass_remove_inplace_ops(graph)
print(graph)    # This looks reasonable too.

torch._C._jit_pass_dce(graph)
print(graph)    # The graph is now empty!
```

If you call `_jit_pass_dce` without first calling `_jit_pass_remove_inplace_ops`, it does **not** produce an empty graph. 

### Versions

/Users/tobyroseman/opt/miniconda3/envs/test/lib/python3.7/site-packages/torch/package/_directory_reader.py:17: UserWarning: Failed to initialize NumPy: numpy.core.multiarray failed to import (Triggered internally at  ../torch/csrc/utils/tensor_numpy.cpp:68.)
  _dtype_to_storage = {data_type(0).dtype: data_type for data_type in _storages}
Collecting environment information...
PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.1 (x86_64)
GCC version: Could not collect
Clang version: 13.0.0 (clang-1300.0.29.3)
CMake version: version 3.12.1
Libc version: N/A

Python version: 3.7.11 (default, Jul 27 2021, 07:03:16)  [Clang 10.0.0 ] (64-bit runtime)
Python platform: Darwin-21.2.0-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] torch==1.10.0
[conda] torch                     1.10.0                   pypi_0    pypi

**Comment by eellison** [comment]
This is a pass maintained by ONNX, not by jit. It shouldn't be in the jit namespace. It does not prove safe semantics. You can run `torch._C._jit_pass_remove_mutation` which will only remove mutation if it can prove it safe semantically. I'm going to close this issue and as a follow up add a warning/comment that this pass doesnt prove safe semantics and to consider the alternative. Please re-open if you have any further thoughts/questions.

**Comment by TobyRoseman** [comment]
@eellison - Thank you for the quick reply. I see what you mean: the graph after calling `_jit_pass_remove_inplace_ops` is just returning the value not writing it. 

It not clear to me why I would use `_jit_pass_remove_mutation` instead of `_jit_pass_remove_inplace_ops`. I can't find any documentation about `_jit_pass_remove_mutation`. Could you explain more about what that method does? Is it related to removing inplace ops?

The GitHub UI will not allow me to reopen this issue. I can only comment. I think this issue should be reopened. This seems like a software regression; it worked correctly in torch version 1.9.1.

**Comment by eellison** [comment]
Here is documentation: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/remove_mutation.h#L68

The reason to `_jit_pass_remove_mutation` instead of `_jit_pass_remove_inplace_ops` would be because it retains mutation when it cannot prove that removing mutation would preserve semantics. 

I'm not sure what the issue should be then - `_jit_pass_dce` is correctly erasing what is dead code that is the result of an incorrect transformation induced by `_jit_pass_remove_inplace_ops`. What does `working correctly` mean here ?

**Comment by TobyRoseman** [comment]
> I'm not sure what the issue should be then - `_jit_pass_dce` is correctly erasing what is dead code that is the result of an incorrect transformation induced by `_jit_pass_remove_inplace_ops`. What does `working correctly` mean here ?

I agree that `_jit_pass_dce` is working correctly. The problem is with `_jit_pass_remove_inplace_ops`. "Working correctly" here means that `_jit_pass_remove_inplace_ops` does not do an incorrect transformation.

Using the example I gave with 1.9.1, the call to `_jit_pass_remove_inplace_ops` does not change the graph. That seems like the correct behavior.

You mentioned that `_jit_pass_remove_inplace_ops` is maintained by ONNX, not jit. So perhaps this is an ONNX issue, but it's certainly an issue which should be addressed.

**Comment by eellison** [comment]
The unfolding of events here is:
1. `_jit_pass_remove_inplace_ops` is introduced which always replaces an inplace op with its functional equivalent, even if that results in a change of semantics, such as the example above, where an input with subsequent uses is now no longer modified

2. - Later, `_jit_remove_mutation` is introduced which will only functionalize ops if we can prove it preserves semantics.

So, the recommendation here as for your issue is just to use `_jit_remove_mutation`. I'm not sure what we should do otherwise - 

we can 
a) make _jit_pass_remove_inplace_ops call into `_jit_remove_mutation`, which will result in better semantics-preservation at the cost of potentially no longer remove mutation on operators which might have actually been valid that we couldn't statically prove.
b) add a warning to `_jit_pass_remove_inplace_ops` about potential loss of semantics and advertise `_jit_pass_remove_mutation`.

I'm wary of doing a) because it might cause export of models which worked previously to break.

**Comment by TobyRoseman** [comment]
> 1. `_jit_pass_remove_inplace_ops` is introduced which always replaces an inplace op with its functional equivalent, even if that results in a change of semantics,

I don't think this is accurate. In torch 1.9.1 the input `graph` has an `aten::copy_` operator which is not replaced when `_jit_pass_remove_inplace_ops` is called.

Not replacing that operator seems like clearly the correct behavior to me.

The input `graph` in torch 1.10.1 does not have a `aten::copy_`. Instead it uses `aten::fill_`. Perhaps `_jit_pass_remove_inplace_ops` has not changed in 1.10.1. Perhaps it just never works for `aten::fill_`.
